core/src/worker/mod.rs

```diff
@@ -791,9 +797,11 @@ impl Worker {
             })
     }
 
-    /// Returns true if shutdown has been requested and there are no more outstanding WFTs
-    fn all_wfts_drained(&self) -> bool {
-        *self.shutdown_requested.borrow() && self.outstanding_workflow_tasks() == 0
+    /// Resolves when shutdown has been requested and there are no more outstanding WFTs
+    async fn all_wfts_drained(&self) {
+        while !(*self.shutdown_requested.borrow() && self.outstanding_workflow_tasks() == 0) {
```

Comment:
You probably don't need `shutdown_requested` here if you're only calling this after a shutdown error.

If you feel like we do need it, it might be better to first wait for shutdown and then wait for no outstanding tasks.